### PR TITLE
Modularize code

### LINKED ISSUE #1: Restructure code into different modules
We started to write our code in the `main.rs`. We should split it into a reasonable module structure.

The following considerations were made:
- Each `struct`, along with its `impl`s and unit-tests should be placed in a separate module
- `main.rs` remains on the top level
- Domain modules should be placed in a common subdirectory

# Acceptance criteria
- Modules are split according to above considerations
- Tests pass

### PR DESCRIPTION
**Pull request by fspiekermann**
implements #1 

Restructures the project in modules:

- every struct has it own module
- parent modules `util` and `order_model` are created
- all tests pass

### CODE REVIEW

File: rusty_pizza_server/src/order_model/meal.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    pub(super) fn new_for_test(meal_id: String, variety: String, price: Money) -> Meal {
+        Meal {
+            meal_id,
+            variety,
+            price,
+            specials: BTreeSet::new(),
+        }
+    }
```
Comment (BjoernLange) [inline rusty_pizza_server/src/order_model/meal.rs:26]: Having this function just to create an instance for tests feels off. "Tests rule" but I think there is a better option. As we already discussed we need to change the way objects are constructed and inserted into the datamodel anyway. In the anticipated version `Meals` would create `Meal`. Therefore there must be a constructor anyway (btw. a [builder](https://doc.rust-lang.org/1.0.0/style/ownership/builders.html) could be a good choice here because specials are optional). So why not make it a `pub` constructor?

File: rusty_pizza_server/src/order_model/meals.rs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    pub(super) fn new_for_test(owner: Rc<User>, ready: bool) -> Meals {
+        Meals {
+            meals: HashMap::new(),
+            owner,
+            ready,
+            paid: Money::new(0, 0),
+            tip: Money::new(0, 0),
+        }
+    }
```
Comment (BjoernLange) [inline rusty_pizza_server/src/order_model/meals.rs]: Same as above.

File: rusty_pizza_server/src/order_model/order.rs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    Ordering,
+    Ordered(String),
+    Delivered,
+}
+
+impl fmt::Display for OrderStatus {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
```
Comment (BjoernLange) [inline rusty_pizza_server/src/order_model/order.rs:20]: Was this added for a reason?

File: rusty_pizza_server/src/order_model/order.rs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        }
+    }
+
+    pub fn add_user(&mut self, user: Rc<User>) {
+        let meals = Meals::new(user.clone());
+        self.meals.insert(user, meals);
+    }
+
+    pub fn get_order_status(self) -> String {
+        self.status.to_string()
```
Comment (BjoernLange) [inline rusty_pizza_server/src/order_model/order.rs]: Why formatting this out as a string?

File: rusty_pizza_server/src/order_model/user.rs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        User { name }
+    }
+
+    pub fn set_name(&mut self, name: String) {
+        self.name = name
+    }
+
+    pub fn get_name(&self) -> String {
+        String::from(self.name.as_str())
+    }
```
Comment (BjoernLange) [inline rusty_pizza_server/src/order_model/user.rs:13]: Why did you add the getter and setter here?

File: rusty_pizza_server/src/order_model/user.rs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    pub fn new(name: String) -> User {
+        User { name }
+    }
+
+    pub fn set_name(&mut self, name: String) {
+        self.name = name
+    }
+
+    pub fn get_name(&self) -> String {
+        String::from(self.name.as_str())
```
Comment (BjoernLange) [inline rusty_pizza_server/src/order_model/user.rs]: I'm asking myself what I would do with the name of the user. Most likely I would simply print it. For that reason a reference `&String` would be sufficient. In case a permanent instance is required I could still clone the value the reference points to. So I would change this to

```suggestion
    pub fn get_name(&self) -> &String {
        &self.name
    }
```

If you really want to copy the name you should use

```suggestion
        self.name.clone()
```

**Review by BjoernLange — CHANGES_REQUESTED** [review]
The new structure looks nice, however I would change the "test constructors" to general constructors. Also, some things were added that are out of the issue's scope (I would still accept it this way, but you should have a look at the comments).

File: rusty_pizza_server/src/order_model/meal.rs (hunk 1 above)
Comment (fspiekermann) [reply rusty_pizza_server/src/order_model/meal.rs:26]: At the end a builder seems the best solution, we should open a Issue for it.
I will change the test constructors to normal constructors in the scope of `pub(super)`. So all models can use these constructors but from outer scope they are not visible

File: rusty_pizza_server/src/order_model/order.rs (hunk 3 above)
Comment (fspiekermann) [reply rusty_pizza_server/src/order_model/order.rs:20]: Yes this is a kind of toString(). To print `Ordered(String)` in prettier way

File: rusty_pizza_server/src/order_model/order.rs (hunk 4 above)
Comment (fspiekermann) [reply rusty_pizza_server/src/order_model/order.rs]: This is a really good question. I think i had an idea during modularize, but at the end this method is not needed anymore.
I will remove it

File: rusty_pizza_server/src/order_model/meal.rs (hunk 1 above)
Comment (fspiekermann) [reply rusty_pizza_server/src/order_model/meal.rs:26]: Since Rust has no method overloading, i combined the constructors to one pub constructor for each model

File: rusty_pizza_server/src/order_model/user.rs (hunk 5 above)
Comment (fspiekermann) [reply rusty_pizza_server/src/order_model/user.rs:13]: Because the name is no longe public, so you won´t be able to access it

Comment (fspiekermann) [reply rusty_pizza_server/src/order_model/user.rs:13]: If we say an user is not able to change his name, we can remove `set_name()`.

Comment (BjoernLange) [reply rusty_pizza_server/src/order_model/user.rs:13]: Yes, as the user is identified by its name making it mutable is a bad idea.

File: rusty_pizza_server/src/order_model/meal.rs (hunk 1 above)
Comment (BjoernLange) [reply rusty_pizza_server/src/order_model/meal.rs:26]: Created #10 for this purpose.

**Review by BjoernLange — APPROVED** [review]
